nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/CaptureChangeMySQL.java

```diff
@@ -0,0 +1,1001 @@
+    private static final SchemaChangeEventWriter schemaChangeEventWriter = new SchemaChangeEventWriter();
+    private static final InsertRowsWriter insertRowsWriter = new InsertRowsWriter();
+    private static final DeleteRowsWriter deleteRowsWriter = new DeleteRowsWriter();
+    private static final UpdateRowsWriter updateRowsWriter = new UpdateRowsWriter();
+
+    static {
+
+        final Set<Relationship> r = new HashSet<>();
+        r.add(REL_SUCCESS);
+        relationships = Collections.unmodifiableSet(r);
+
+        final List<PropertyDescriptor> pds = new ArrayList<>();
+        pds.add(HOSTS);
+        pds.add(DRIVER_NAME);
+        pds.add(DRIVER_LOCATION);
+        pds.add(USERNAME);
+        pds.add(PASSWORD);
+        pds.add(DATABASE_NAME_PATTERN);
+        pds.add(TABLE_NAME_PATTERN);
+        pds.add(CONNECT_TIMEOUT);
+        pds.add(DIST_CACHE_CLIENT);
+        pds.add(RETRIEVE_ALL_RECORDS);
+        pds.add(STATE_UPDATE_INTERVAL);
+        pds.add(INIT_SEQUENCE_ID);
+        pds.add(INIT_BINLOG_FILENAME);
+        pds.add(INIT_BINLOG_POSITION);
+        propDescriptors = Collections.unmodifiableList(pds);
+    }
+
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return propDescriptors;
+    }
+
+    public void setup(ProcessContext context) {
+
+        final ComponentLog logger = getLogger();
+
+        final StateManager stateManager = context.getStateManager();
+        final StateMap stateMap;
+
+        try {
+            stateMap = stateManager.getState(Scope.CLUSTER);
+        } catch (final IOException ioe) {
+            logger.error("Failed to retrieve observed maximum values from the State Manager. Will not attempt "
+                    + "connection until this is accomplished.", ioe);
+            context.yield();
+            return;
+        }
+
+        Map<String, String> statePropertiesMap = new HashMap<>(stateMap.toMap());
+
+        PropertyValue dbNameValue = context.getProperty(DATABASE_NAME_PATTERN);
+        databaseNamePattern = dbNameValue.isSet() ? Pattern.compile(dbNameValue.getValue()) : null;
+
+        PropertyValue tableNameValue = context.getProperty(TABLE_NAME_PATTERN);
+        tableNamePattern = tableNameValue.isSet() ? Pattern.compile(tableNameValue.getValue()) : null;
+
+        stateUpdateInterval = context.getProperty(STATE_UPDATE_INTERVAL).evaluateAttributeExpressions().asTimePeriod(TimeUnit.MILLISECONDS);
+
+        // Pre-store initial sequence ID if none has been set in the state map
+        final PropertyValue initSequenceId = context.getProperty(INIT_SEQUENCE_ID);
+        statePropertiesMap.computeIfAbsent(EventWriter.SEQUENCE_ID_KEY, k -> initSequenceId.isSet() ? initSequenceId.evaluateAttributeExpressions().getValue() : "0");
```

Comment:
`statePropertiesMap` is derived from stateMap.toMap and it's a different map instance. Adding new entry to `statePropertiesMap` doesn't affect stateMap. And `statePropertiesMap` is not used by any further code but here. I wonder if this is expected. Is it correct?